cmd/object-handlers.go

```diff
@@ -1672,6 +1677,361 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 	})
 }
 
+type tarObjectData struct {
+	Data  []byte
+	Name  string
+	Info  os.FileInfo
+	Error error
+}
+
+func extractTARGZ(r io.Reader) (<-chan tarObjectData, error) {
+	gzf, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+
+	tarReader := tar.NewReader(gzf)
+	chReader := make(chan tarObjectData)
+	go func() {
+		defer close(chReader)
+		bb := bytebufferpool.Get()
+		defer func() {
+			bb.Reset()
+			bytebufferpool.Put(bb)
+		}()
+		for {
+			header, err := tarReader.Next()
+			if err != nil {
+				if err != io.EOF {
+					logger.LogIf(GlobalContext, err)
+					chReader <- tarObjectData{
+						Error: err,
+					}
+				}
+				return
+			}
+
+			name := header.Name
+			if name == slashSeparator {
+				continue
+			}
+
+			switch header.Typeflag {
+			case tar.TypeDir: // = directory
+				chReader <- tarObjectData{
+					Data: nil,
+					Name: trimLeadingSlash(pathJoin(name, slashSeparator)),
+					Info: header.FileInfo(),
+				}
+			case tar.TypeReg: // = regular
+				if _, err = io.CopyN(bb, tarReader, header.Size); err != nil {
```

Comment:
This should be a streaming operation so you don't risk gzip bombs filling up memory. The first thing you do when you receive it is to convert it to a reader anyway.

It would require you to wait for the upload to finish (probably ditching channels, just use a callback), but that is no big change anyway.

Reply:
@harshavardhana  Send the `tarReader` as replacement for `reader = bytes.NewReader(obj.Data)` below.

As mentioned you need it to finish before you can move on to the next object, therefore a callback is probably better than a channel, or you will have to send back the state on the channel.